scipy/stats/_continuous_distns.py

```diff
@@ -3554,6 +3559,66 @@ def fit(self, data, *args, **kwds):
             fshape_s = fshape_n / fscale
         return fshape_s, floc, fscale
 
+    def _ppf_isf(self, q, mu, upper):
+        """ calculate ppf or isf depending on the value of `upper`.
+        If upper is True, then compute isf, else ppf. This implementation is
+        based on Gyner & Smyth (2016) and uses newton's method with carefully
+        selected starting points that guarantee convergence regardless of the
+        probability."""
+
+        def f(a, b, c, d, e):
+            """approximates (x0 - norm_cdf) / pdf when abs(x0 - norm_cdf) < 1e-5"""
+            return a * np.exp(c + np.log1p(-0.5 * a) - d)
+
+        def f2(a, b, c, d, e):
+            """computes (x0 - norm_cdf) / pdf"""
+            return b * np.exp(-d) - np.exp(e - d)
+
+        def f_small_q(q, mu, k):
+            """select the start values for the newton using the quantiles of
+            the gamma (if right tail is required) or normal distribution"""
+            if upper:
+                mu3 = mu ** 3
+                return sc.gammainccinv(1 / mu3, q) / mu3
+            return mu / (_norm_ppf(q) ** 2),
+
+        k = 1.5 * mu
+        # get starting points
+        x0 = _lazywhere(
+            q < 1e-5,
+            (q, mu, k),
+            f_small_q,
+            f2=lambda q, mu, k: _lazywhere(
+                mu > 100,
+                (mu, k),
+                lambda mu, k: mu * (0.5 / k - 0.125 / (k ** 3) + 0.0625 / (k ** 6)),
+                f2=lambda mu, k: mu * (np.sqrt(1 + k * k) - k)
+            )
+        )
+
+        lq = np.log(q)
+        iterations = 0
+        lx_func = self._logsf if upper else self._logcdf
+        sign = -1 if upper else 1
+        while iterations < 100:
+            lx = lx_func(x0, mu)
+            d = lq - lx
+            lp = self._logpdf(x0, mu)
+            delta = _lazywhere(np.abs(d) < 1e-5, (d, q, lq, lp, lx), f, f2=f2)
```

Comment:
Same as above. Much cleaner if you do two separate liners. Then you don't need _lazywhere

Reply:
Doesn't that mean both `f` and `f2` get evaluated even when all elements of d are greater than 1e-05?. I assumed that _lazywhere is to prevent such cases. Or am I misinterpreting your comment?

Reply:
I mean something like; first get the condition elements

```python
ind = np.abs(d) < 1e-5
delta[ind] = ... f related stuff
delta[~ind] = ... f2 related stuff
```

Reply:
That's what I wanted to confirm. If `~ind` is all False values then doesn't that mean `f2 related stuff` gets executed anyway?

Reply:
Ideally it shouldn't because the right hand side would also depend on `d[~ind]` which should have no elements either. Hence it would mostly be empty. You can explicitly test for it anyways. We don't need to do code golfing as long as it is clear.

Reply:
what happens when inputs are scalars? using [] indexing wont work. I am getting constant failures on tests because this approach requires explicit handling of several cases where input has different dimensions. using _lazywhere handles the size and shape differences internally. Do you have any suggestions? I am thinking that adding additional checks on top of splitting the lines might not improve readability here.